rust/src/transformations/make_stable_expr/expr_clip/mod.rs

```diff
@@ -0,0 +1,158 @@
+            DataType::UInt64 => extract_bounds::<u64>(lower, upper, series_domain)?,
+            DataType::Int8 => extract_bounds::<i8>(lower, upper, series_domain)?,
+            DataType::Int16 => extract_bounds::<i16>(lower, upper, series_domain)?,
+            DataType::Int32 => extract_bounds::<i32>(lower, upper, series_domain)?,
+            DataType::Int64 => extract_bounds::<i64>(lower, upper, series_domain)?,
+            DataType::Float32 => extract_bounds::<f32>(lower, upper, series_domain)?,
+            DataType::Float64 => extract_bounds::<f64>(lower, upper, series_domain)?,
+            _ => return fallible!(MakeTransformation, "unsupported dtype"),
+        }
+    };
+
+    let series_name = input_domain.active_series()?.field.name.as_str();
+
+    // Margins on the active_column could be preserved but this functionality has not been implemented yet
+    output_domain.frame_domain.margins = output_domain
+        .frame_domain
+        .margins
+        .into_iter()
+        .filter(|(s, _)| !s.contains(series_name))
+        .collect();
+
+    t_prior
+        >> Transformation::new(
+            input_domain.clone(),
+            output_domain,
+            Function::new_expr(move |expr| expr.clip(lower.clone(), upper.clone())),
+            input_metric.clone(),
+            input_metric,
+            StabilityMap::new(Clone::clone),
+        )?
+}
+
+fn extract_bound<T: NumericDataType>(bound: Expr) -> Fallible<T> {
+    let Expr::Literal(value) = bound else {
+        return fallible!(MakeTransformation, "bound must be a literal");
+    };
+    let value = value
+        .to_any_value()
+        .ok_or_else(|| err!(MakeTransformation, "bound must be a numeric dtype"))?;
+
+    value.try_extract().map_err(|_| {
+        err!(
+            MakeTransformation,
+            "bound cannot be represented in the expected type"
+        )
+    })
+}
+
+fn extract_bounds<T: Number + NumericDataType + Literal>(
+    lower: Expr,
+    upper: Expr,
+    domain: &mut SeriesDomain,
+) -> Fallible<(Expr, Expr)> {
+    let bounds = (extract_bound::<T>(lower)?, extract_bound::<T>(upper)?);
+
+    let mut atom_domain = domain
+        .element_domain
+        .as_any()
+        .downcast_ref::<AtomDomain<T>>()
+        .ok_or_else(|| err!(MakeTransformation, "lower bound must be a literal"))?
+        .clone();
+
+    atom_domain.bounds = Some(Bounds::new_closed(bounds)?);
+    domain.element_domain = Arc::new(atom_domain);
+
+    Ok((bounds.0.lit(), bounds.1.lit()))
+}
+
+#[cfg(test)]
+mod test_make_clamp {
+    use crate::domains::{AtomDomain, LogicalPlanDomain};
+    use crate::metrics::SymmetricDistance;
+    use crate::transformations::polars_test::get_test_data;
+
+    use super::*;
+
+    #[test]
+    fn test_make_clip_expr() -> Fallible<()> {
+        let (lf_domain, lf) = get_test_data()?;
+        let lp = lf.logical_plan.clone();
+        let expr_domain = lf_domain.clone().select();
```

Comment:
This might be Rust basics, but what goes wrong if you don't clone?

Reply:
`lf.logical_plan` didn't need to be cloned. The rust compiler considers this case a partial move-- `lp` now owns the memory behind the `logical_plan` field in the `lf` struct. If some code afterwards tried to use `lf.logical_plan`, or even `lf`, the compiler would complain. But in this code, nothing used it.

`lf_domain` does need to be cloned, though. The signature of `select` is `fn select(self) -> ExprDomain`, meaning it takes ownership of `self` (in this case, `self` is `lf_domain.clone()`). If there were no clone, then the use of `lf_domain` later in this function would result in a compiler error. That memory would now owned by `expr_domain`.